pyamg/amg_core/ruge_stuben.h

```diff
@@ -956,345 +943,1339 @@ void cr_helper(const I A_rowptr[],  const int A_rowptr_size,
+ *
+ * Returns:
+ * --------
+ * Nothing, Pj[] and Px[] modified in place.
+ *
+ * Notes:
+ * ------
+ * Includes connections a_ji from j to point i itself in interpolation weights
+ * to improve interpolation.
+ *
+ * References:
+ * -----------
+ * [0] "Distance-Two Interpolation for Parallel Algebraic Multigrid,"
+ *      H. De Sterck, R. Falgout, J. Nolting, U. M. Yang, (2008).
+ */
+template<class I, class T>
+void extended_plusi_interpolation_pass2(const I n_nodes,
+                                        const I Ap[], const int Ap_size,
+                                        const I Aj[], const int Aj_size,
+                                        const T Ax[], const int Ax_size,
+                                        const I Sp[], const int Sp_size,
+                                        const I Sj[], const int Sj_size,
+                                        const T Sx[], const int Sx_size,
+                                        const I splitting[], const int splitting_size,
+                                        const I Pp[], const int Pp_size,
+                                              I Pj[], const int Pj_size,
+                                              T Px[], const int Px_size)
+{
+
+    for (I i = 0; i < n_nodes; i++) {
+        // If node i is a C-point, then set interpolation as injection
+        if(splitting[i] == C_NODE) {
+            Pj[Pp[i]] = i;
+            Px[Pp[i]] = 1;
+        } 
+        // Otherwise, use extended+i distance-two AMG interpolation formula
+        // (see Eqs. 4.10-4.11 in [0])
+        else {
+
+            // -------------------------------------------------------------------------------- //
+            // -------------------------------------------------------------------------------- //
+            // Calculate outer denominator
+            T denominator = 0.0;
+
+            // Start by summing entire row of A
+            for (I mm = Ap[i]; mm < Ap[i+1]; mm++) {
+                denominator += Ax[mm];
+            }
+
+            // Then subtract off the strong connections so that you are left with 
+            // denominator = a_ii + sum_{m in weak connections} a_im
+            for (I mm = Sp[i]; mm < Sp[i+1]; mm++) {
+                I this_point = Sj[mm];
+                if ( this_point != i ) {
+                    denominator -= Sx[mm]; // making sure to leave the diagonal entry in there
+                }
+                // Subtract off distance-two strong C connections that are also distance-one weak
+                // connections (i.e. not yet removed from the weak connections in denominator)
+                if (splitting[this_point] == F_NODE && (this_point != i)) {
+
+                    for (I ff = Sp[this_point]; ff < Sp[this_point+1]; ff++) {
+                        I d2_point = Sj[ff];
+
+                        // Strong C-connections to strong F-connections (distance two C connections)
+                        if (splitting[d2_point] == C_NODE) {
+
```

Comment:
This may also be a bug.  I've been studying equation 20 in the paper.  You should only subtract Ax[search_ind] from the denominator if you know that it Aj[search_ind] is a weak connection.  Do we know that?  

I don't think so (but I could be wrong, this stuff is hard!).

Reply:
So here's a first question - should the N_i^w really be \hat{N}_i^w? Paper just says N_i^w, i.e., weak connections of distance one, but this seems weird given the distance two nature of the interpolation and equation (13) for standard.

Reply:
I think you're right.  If it's N_i^w and not \hat{N}_i^w, then there are no distance-two connections in N.  And, this doesn't make much sense.  
I also still suspect that the loop `for (I search_in = Ap[i]; ...)` below, should actually be over only weak connections.